File: decorator/decorator.py
class Coffee:
    def cost(self):
        return 5

class CoffeeDecorator(Coffee):
    def __init__(self, coffee):
        self._coffee = coffee
    
    def cost(self):
        return self._coffee.cost()

class MilkDecorator(CoffeeDecorator):
    def cost(self):
        return self._coffee.cost() + 2

class SugarDecorator(CoffeeDecorator):
    def cost(self):
        return self._coffee.cost() + 1

coffee = Coffee()
coffee.cost() # 5

coffee_with_milk = MilkDecorator(coffee)
coffee_with_milk.cost() # 7

coffee_with_sugar = SugarDecorator(coffee)
coffee_with_sugar.cost() # 6

coffee_with_milk_and_sugar = SugarDecorator(coffee_with_milk)
coffee_with_milk_and_sugar.cost() # 8

File: decorator/decorator.kt
open class Coffee {
    open fun cost(): Int {
        return 5
    }
}

open class CoffeeDecorator(private val coffee: Coffee) : Coffee() {
    override fun cost(): Int {
        return coffee.cost()
    }
}

class MilkDecorator(coffee: Coffee) : CoffeeDecorator(coffee) {
    override fun cost(): Int {
        return super.cost() + 2
    }
}

class SugarDecorator(coffee: Coffee) : CoffeeDecorator(coffee) {
    override fun cost(): Int {
        return super.cost() + 1
    }
}

fun main() {
    val coffee = Coffee()
    println("Plain Coffee: ${coffee.cost()}")

    val coffeeWithMilk = MilkDecorator(coffee)
    println("Coffee with Milk: ${coffeeWithMilk.cost()}")

    val coffeeWithSugar = SugarDecorator(coffee)
    println("Coffee with Sugar: ${coffeeWithSugar.cost()}")

    val coffeeWithMilkAndSugar = SugarDecorator(coffeeWithMilk)
    println("Coffee with Milk and Sugar: ${coffeeWithMilkAndSugar.cost()}")
}